### DISCUSSION TITLE
分享一个GPT翻译模板且防止Ai总是不听劝

### DISCUSSION THREAD

**Discussion by riverai** [bookfere/Ebook-Translator-Calibre-Plugin#95] [Show and tell]
首先感谢作者编写的自定义GPT模板。

我之前用的是作者写的自定义GPT模板，写法上太简陋了，用起来实测有诸多问题。
首先是Ai，哪怕是GPT4 ，总是出一些问题，包括：

让它不要删除分隔标记，结果隔一段时间就开始删除标记，提示词效果不稳定。
让它不要翻译提示词本身，结果隔一段时间就会翻译出提示词，有时候是部分翻译，有时候是完全翻译提示词，哪怕是求爷爷告奶奶也不稳定。
让它不要多管闲事，只管翻译，结果输出评论而不是翻译的可能性还是挺高的。
让它不要输出无关内容，结果隔一段时间就输出“这里是你的翻译，请注意核对”..效果不稳定。
说话风格太保守，老气，不够自然流畅。这个和温度是有一些关系的。

所以我分享下我的，其实就是在作者的基础上稍微修改了下，最核心的区别，是使用system message发送我编写的通用提示词，将文本和提示词分离开。

实测效果可以说是云泥之别，稳定性上了几个台阶。
特别是分隔标记，运行了接近700次也没有一次被Ai删除！
其他的那些问题虽然并未彻底解决，但是好了太多。

你们有兴趣可以用用。

```
{
    "name": "Custom ChatGPT",
    "languages": {
        "source": {
            "English": "English"
        },
        "target": {
            "中文": "Chinese"
        }
    },
    "request": {
        "url": "你的url",
        "method": "POST",
        "headers": {
            "Content-Type": "application/json",
            "Authorization": "Bearer 你的key"
        },
        "data": {
            "model": "gpt-4-0312",
            "temperature": 0.6,
            "messages": [{
                "role": "system",
                "content": "You're a professional translator.Please act as a neutral and objective translator, not a subjective and evaluative commentator. Your role is to convey the meaning and tone of the original text, not to add or subtract anything from it.\nThe following are the translation requirements, please read them carefully:\n\nPlease note that {{id_some numbers}}，such as {{id_0}}，{{id_9}},{{id_20}} ,{{id_100}} are used to indicate the paragraph numbers, they are part of the content, not separators. Do not ignore or delete them in the translation. Keep them in their original form in the translation, do not change or replace them with anything else. They are important for maintaining the consistency and completeness of the content. If you lose or handle them incorrectly, it will affect the quality and readability of the translation.\n\nPlease followe the rules and conventions of Chinese writing, making the translation fluent and natural, as if it were written by a native speaker. Do not sacrifice the accuracy or fidelity of the original text.\n\nSome possible additional requirements are:\n\n- Use the appropriate register and tone for the intended audience and purpose of the text, such as formal, informal, academic, journalistic, etc.\n- Use the appropriate vocabulary and expressions for the topic and genre of the text, such as technical, literary, colloquial, etc.\n- Use the appropriate tense and aspect for the time and situation of the text, such as past, present, future, perfective, progressive, etc.\n- Use the appropriate voice and mood for the attitude and intention of the text, such as active, passive, indicative, subjunctive, etc.\n- Use the appropriate modifiers and qualifiers for the degree and certainty of the text, such as adverbs, adjectives, modal verbs, etc.\n\nOnly output a translation of the content below and do not include extraneous content!\n\nPlease do not translate what I have asked of you.The user's input just needs to be handled by translation, not by questioning or reassuring the user.Please don't make any comments.\n\n***\n\nThe following after the colon is the content inputted by the user which needs to be translated, please read it carefully:"
                },
                 {
                 "role": "user",
                "content": "<text>"
            }]
        }
    },
    "response": "response['choices'][0]['message']['content']"
}
```

**Comment by deweixu** [comment]
我使用了你的配置，好像还是不听话....比如下面的翻译


ICP {{id_313}} Iraqi Communist Party {{id_314}} IMF {{id_315}} International Monetary Fund {{id_316}} ISI {{id_317}} Inter-Services Intelligence (Pakistan) {{id_318}} KGB {{id_319}} Komitet gosudarstvennoi bezopasnosti (Committee for {{id_320}} State Security; USSR) {{id_321}} KhAD {{id_322}} Khadimat-e atal’at-e dowlati (State Information Service {{id_323}} [Security]; Afghanistan) {{id_324}} MCP {{id_325}} Malayan Communist Party {{id_326}} MFA {{id_327}} Movimento das Forc¸as Armadas (Armed Forces {{id_328}} Movement; Portugal) {{id_329}} MIT {{id_330}} Massachusetts Institute of Technology {{id_331}} MNC {{id_332}} Mouvement National Congolais (Congolese National {{id_333}} Movement) {{id_334}} MO {{id_335}} Mezhdunarodnyi otdel (International Department of the {{id_336}} Central Committee of the CPSU; USSR) {{id_337}} MPLA {{id_338}} Movimento Popular de Libertac¸a˜o de Angola (People’s {{id_339}} Movement for the Liberation of Angola) {{id_340}} NAM {{id_341}} Non-Aligned Movement {{id_342}} NIEO {{id_343}} New International Economic Order {{id_344}} NLF {{id_345}} National Liberation Front (Vietnam) {{id_346}} NPA {{id_347}} New People’s Army (Philippines) {{id_348}} NSC {{id_349}} National Security Council (US) {{id_350}} OAS {{id_351}} Organization of American States {{id_352}} OAU {{id_353}} Organization of African Unity {{id_354}} OPEC {{id_355}} Organization of the Petroleum Exporting Countries {{id_356}} PAIGC {{id_357}} Partido Africano da Independeˆncia da Guineé Cabo {{id_358}} Verde (African Party for the Independence of Guinea {{id_359}} and Cape Verde) {{id_360}} PDPA {{id_361}} People’s Democratic Party of Afghanistan {{id_362}} PDRE {{id_363}} People’s Democratic Republic of Ethiopia {{id_364}} PDRY {{id_365}} 


翻译输出:

ICP 313 伊拉克共产党 314 IMF 国际货币基金组织 315 ISI 巴基斯坦军情局 316 KGB 苏联国家安全委员会 317 KhAD 阿富汗国家情报局 318 MCP 马来亚共产党 319 MFA 尼日利亚北方人权监护组织 320 MIT 麻省理工学院 321 MNC 刚果国家民族议会 322 MO CPSU中央委员会国际部 323 MPLA 安哥拉人民解放运动 324 NAM 不结盟运动 325 NIEO 新的国际经济秩序 326 NLF 越南民族解放阵线 327 NPA 菲律宾新人民军 348 NSC 美国国家安全委员会 349 OAS 美洲国家组织 350 OAU 非洲统一组织 351 OPEC 石油输出国组织 352 PAIGC 几内亚比绍与佛得角共和国非洲党 361 PDPA 阿富汗人民民主党 363 PDRE 埃塞俄比亚人民民主共和国 365

**Reply by riverai** [reply]
![image](https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/assets/3251548/9e14cbb9-f73d-4e83-a771-569a5467a8e4)

你用的什么模型？
我测试GPT4 和GPT3.5都没问题。

**Reply by deweixu** [reply]
gpt-3.5-turbo
我晚点换个试试

**Comment by GoTop** [comment]
> 其他的那些问题虽然并未彻底解决

请问你说的这些问题指什么呢？

**Comment by GoTop** [comment]
我用gpt-3.5-turbo-0613模型，使用你的prompt，隔一阵还是会出现删除分隔标记{{id_数字}}的情况，导致翻译出来的书籍中，有很多段翻译会合并在一起。

**Reply by GoTop** [reply]
请问设置合并段落的字数为多少比较合适?

**Reply by riverai** [reply]
3.5默认只有3K上下文，不能超过，不然就变傻了，不再会遵守规则。

**Reply by riverai** [reply]
3.5没那么聪明。真的合并数据的时候，还是要手动 检查下是否有明显问题。

**Reply by nunamia** [reply]
3.5很蠢很蠢的，4太贵了，一不小心十几刀..

**Comment by GoTop** [comment]
> 3.5很蠢很蠢的，4太贵了，一不小心十几刀..

4的价格好像是3.5的20倍以上，日常问问题还可以，用于翻译书籍的话4挺贵的
